File: solutions/javascript/clock/1/clock.js
//
// This is only a SKELETON file for the 'Clock' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class Clock {
  constructor(hour, minute = 0) {
    this.totalMinutes = this._normalize(hour, minute);
  }

  _normalize(hour, minute) {
    const total = hour * 60 + minute;
    const minutesPerDay = 24 * 60;
    return ((total % minutesPerDay) + minutesPerDay) % minutesPerDay;
  }

  toString() {
    const hours = Math.floor(this.totalMinutes / 60);
    const minutes = this.totalMinutes % 60;
    return `${hours.toString().padStart(2, '0')}:${minutes.toString().padStart(2, '0')}`;
  }

  plus(minutes) {
    return new Clock(0, this.totalMinutes + minutes);
  }

  minus(minutes) {
    return this.plus(-minutes);
  }

  equals(other) {
    return this.totalMinutes === other.totalMinutes;
  }
}

File: solutions/cpp/clock/1/clock.cpp
#include "clock.h"

#include <iomanip>
#include <sstream>

namespace date_independent {

clock clock::at(int hours, int minutes) {
    return clock(hours, minutes);
}

clock::clock(int hours, int minutes) {
    int total = hours * 60 + minutes;
    // Normalize to [0, 1440)
    total %= 1440;
    if (total < 0) {
        total += 1440;
    }
    minutes_since_midnight_ = total;
}

clock clock::plus(int minutes) const {
    return clock(0, minutes_since_midnight_ + minutes);
}

clock clock::minus(int minutes) const {
    return clock(0, minutes_since_midnight_ - minutes);
}

bool clock::operator==(const clock& other) const {
    return minutes_since_midnight_ == other.minutes_since_midnight_;
}

bool clock::operator!=(const clock& other) const {
    return !(*this == other);
}

std::string clock::to_string() const {
    int hours = minutes_since_midnight_ / 60;
    int mins = minutes_since_midnight_ % 60;
    std::ostringstream oss;
    oss << std::setw(2) << std::setfill('0') << hours << ':'
        << std::setw(2) << std::setfill('0') << mins;
    return oss.str();
}

clock::operator std::string() const {
    return to_string();
}

}  // namespace date_independent